File: backend/internal/models/claim.go
package models

import (
	"database/sql/driver"
	"encoding/json"
	"fmt"
	"time"
)

// IntArray is a custom type for handling JSONB integer arrays
type IntArray []int

// Scan implements the sql.Scanner interface for IntArray
func (a *IntArray) Scan(value interface{}) error {
	if value == nil {
		*a = []int{}
		return nil
	}

	bytes, ok := value.([]byte)
	if !ok {
		return fmt.Errorf("failed to scan IntArray: expected []byte, got %T", value)
	}

	return json.Unmarshal(bytes, a)
}

// Value implements the driver.Valuer interface for IntArray
func (a IntArray) Value() (driver.Value, error) {
	if a == nil {
		return []byte("[]"), nil
	}
	return json.Marshal(a)
}

type Claim struct {
	ID              string     `json:"id" db:"id"`
	PropertyID      string     `json:"property_id" db:"property_id"`
	PolicyID        string     `json:"policy_id" db:"policy_id"`
	ClaimNumber     *string    `json:"claim_number" db:"claim_number"`
	LossType        string     `json:"loss_type" db:"loss_type"`
	IncidentDate    time.Time  `json:"incident_date" db:"incident_date"`
	Status          string     `json:"status" db:"status"`
	FiledAt         *time.Time `json:"filed_at" db:"filed_at"`
	Description     *string    `json:"description" db:"description"`

	// Step tracking
	CurrentStep    int      `json:"current_step" db:"current_step"`
	StepsCompleted IntArray `json:"steps_completed" db:"steps_completed"`

	// Step-specific fields
	ContractorEmail            *string    `json:"contractor_email" db:"contractor_email"`
	ContractorName             *string    `json:"contractor_name" db:"contractor_name"`
	ContractorPhotosUploadedAt *time.Time `json:"contractor_photos_uploaded_at" db:"contractor_photos_uploaded_at"`
	ContractorEstimateTotal    *float64   `json:"contractor_estimate_total" db:"contractor_estimate_total"`
	DeductibleComparisonResult *string    `json:"deductible_comparison_result" db:"deductible_comparison_result"`
	InsuranceClaimNumber       *string    `json:"insurance_claim_number" db:"insurance_claim_number"`
	InspectionDatetime         *time.Time `json:"inspection_datetime" db:"inspection_datetime"`

	// Existing fields
	AssignedUserID  *string    `json:"assigned_user_id" db:"assigned_user_id"`
	AdjusterName    *string    `json:"adjuster_name" db:"adjuster_name"`
	AdjusterPhone   *string    `json:"adjuster_phone" db:"adjuster_phone"`
	MeetingDatetime *time.Time `json:"meeting_datetime" db:"meeting_datetime"`
	CreatedByUserID string     `json:"created_by_user_id" db:"created_by_user_id"`
	CreatedAt       time.Time  `json:"created_at" db:"created_at"`
	UpdatedAt       time.Time  `json:"updated_at" db:"updated_at"`
}

type ClaimActivity struct {
	ID           string     `json:"id" db:"id"`
	ClaimID      string     `json:"claim_id" db:"claim_id"`
	UserID       *string    `json:"user_id" db:"user_id"`
	ActivityType string     `json:"activity_type" db:"activity_type"`
	Description  string     `json:"description" db:"description"`
	Metadata     *string    `json:"metadata" db:"metadata"` // JSON string
	CreatedAt    time.Time  `json:"created_at" db:"created_at"`
}

File: frontend/src/types/claim.ts
export type LossType = 'water' | 'hail'

// ─── PM Brain Strategy Engine ──────────────────────────────────────────────
export type PMBrainStatus = 'CLOSE' | 'DISPUTE_OFFER' | 'LEGAL_REVIEW' | 'NEED_DOCS'

export interface PMBrainDeltaDriver {
  line_item: string
  contractor_price: number
  carrier_price: number
  delta: number
  reason: string
}

export interface PMBrainCoverageDispute {
  item: string
  status: 'denied' | 'partial'
  contractor_position: string
}

export interface PMBrainAnalysis {
  status: PMBrainStatus
  plain_english_summary: string
  total_contractor_estimate: number
  total_carrier_estimate: number
  total_delta: number
  top_delta_drivers: PMBrainDeltaDriver[]
  coverage_disputes: PMBrainCoverageDispute[]
  required_next_steps: string[]
  legal_threshold_met: boolean
}
export type ClaimStep = 1 | 2 | 3 | 4 | 5 | 6 | 7
export type DeductibleResult = 'worth_filing' | 'not_worth_filing'
export type ViabilityRecommendation = 'PURSUE' | 'PURSUE_WITH_CONDITIONS' | 'DO_NOT_PURSUE'

export interface ViabilityAnalysis {
  recommendation: ViabilityRecommendation
  net_estimated_recovery: number
  coverage_score: number
  economics_score: number
  top_risks: string[]
  required_next_steps: string[]
  plain_english_summary: string
}

export interface Payment {
  id: string
  claim_id: string
  payment_type: 'acv' | 'rcv'
  amount: number
  expected_amount?: number
  check_number?: string
  received_date?: string
  notes?: string
  status: 'expected' | 'received' | 'reconciled' | 'disputed'
  created_at: string
  updated_at: string
}

export interface Property {
  id: string
  organization_id: string
  nickname: string
  legal_address: string
  lat?: number
  lng?: number
  owner_entity_name: string
  mortgage_bank_id?: string
  status: string
  created_at: string
  updated_at: string
  policy?: {
    policy_number: string
    carrier: string
  }
}

export interface Policy {
  id: string
  property_id: string
  carrier_name: string
  carrier_phone?: string | null
  carrier_email?: string | null
  policy_number: string
  deductible_value: number
  exclusions?: string | null
  policy_pdf_url?: string | null
  effective_date?: string | null
  expiration_date?: string | null
  created_at: string
  updated_at: string
}

export interface Claim {
  id: string
  claim_number: string | null
  property_id: string
  property?: Property
  policy?: Policy
  loss_type: LossType
  status: string
  incident_date: string
  filed_at?: string
  description?: string
  contractor_estimate_total?: number

  // Step tracking
  current_step?: ClaimStep
  steps_completed?: number[]

  // Step-specific data
  contractor_email?: string
  contractor_name?: string
  contractor_photos_uploaded_at?: string
  deductible_comparison_result?: DeductibleResult
  insurance_claim_number?: string
  adjuster_name?: string
  adjuster_phone?: string
  inspection_datetime?: string

  created_at: string
  updated_at: string

  // Legal escalation (added in migration 000011)
  legal_partner_name?: string
  legal_partner_email?: string
  owner_email?: string
  legal_escalation_status?: 'pending_approval' | 'approved' | 'declined' | 'sent_to_lawyer'
}
